[Security] reusing AuthorizationChecker in AccessListener

oleg-andreyev opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Fixed tickets | #21571

While investigating why `AuthenticationEvents::AUTHENTICATION_SUCCESS` is not dispatched, noticed similarity between `AccessListener` and `AuthorizationChecker` and the only real difference is `alwaysAuthenticate` flag.

**Case 1:**
When `UsernamePasswordToken` is serialized it will be serialized with state of `authenticated`,
so when application will load token from session `AccessListener` won't call `authenticate` because `authenticated=true`, even if `alwaysAuthenticate=true`

**Case 2:**
User A has roles `foo, bar and admin`, User A is signed-in into application and token is persisted, later another User B with role `admin`, decided to restrict role `admin` for User A, so User A won't lose it's privileges until session is expired or logout, because token is persisted with `roles` and `authenticated=true` and AccessListener does not call `authenticate` even if `alwaysAuthenticate=true`

**TODO**
- [x] - reuse AuthorizationChecker in AccessListener
- [ ] - adjust unit tests
- [x] - add deprecation notices
- [x] - update CHANGELOG and UPGRADE*

iltar commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -26,17 +24,13 @@
+    private $authorizationChecker;
 
-    public function __construct(TokenStorageInterface $tokenStorage, AccessDecisionManagerInterface $accessDecisionManager, AccessMapInterface $map, AuthenticationManagerInterface $authManager)
+    public function __construct(AuthorizationCheckerInterface $authorizationChecker, AccessMapInterface $map)
```
This is a BC break, so it can't be merged as is. Not sure if BC breaks should be added in older versions or if we should not see this class as an extension point (which it is atm).

stof replied:
even then, we still need to preserve BC even in master

oleg-andreyev replied:
sure, I'll prepare BC changes in near time.

iltar commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -24,23 +27,86 @@
+     * @var TokenStorageInterface
+     * @deprecated
+     */
+    private $tokenStorage;
```
The docblocks and deprecation notices are not required for private properties (if the type is known). Private means it's not an extension point, thus don't have to marked as such.

oleg-andreyev replied:
removed docblock. thanks.

iltar commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -24,23 +27,86 @@
+        }
+
+        if ($oldSignature) {
+            @trigger_error('The '.__CLASS__.' constructor now expects 2 arguments', E_USER_DEPRECATED);
```
This message should mentioned what's deprecated as of when and how to fix it

oleg-andreyev replied:
Updated deprecation message, but I would like to get suggestions on better wording

oleg-andreyev commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -30,22 +31,85 @@ class AccessListener implements ListenerInterface
+        }
+
+        if ($oldSignature) {
+            @trigger_error('The '.__CLASS__.' constructor expects 2 arguments since version X.X. Update constructor signature AccessListener(AuthorizationCheckerInterface, AccessMapInterface).', E_USER_DEPRECATED);
```
`X.X` will be replaced with correspond version as soon as decision will be made where this code will land.

oleg-andreyev commented:
@nicolas-grekas @iltar @stof can you give an advice against what branch this changes should go?
So 2.7 version entered security fixes phase 2 days ago.

iltar commented:
My suggestion would be to fix this in 2.7 if it's a bug (which it does sound like), and then a feature for 4.2 to fix the constructor, making it an extra optional argument until the change of signature.

oleg-andreyev commented:
Ok, I'll prepare my PR for 2.7 as bugfix, and then will prepare PR with deprecation for master

oleg-andreyev commented:
@iltar @stof @nicolas-grekas all set and could be reviewed.

iltar commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml:
```diff
@@ -266,6 +266,7 @@
             <argument type="service" id="security.access.decision_manager" />
             <argument type="service" id="security.access_map" />
             <argument type="service" id="security.authentication.manager" />
+            <argument type="service" id="security.authorization_checker" on-invalid="null"/>
```
I'm not entirely sure the `on-invalid="null"` is needed. In case of an additional argument missing (old component version), it would just pass an extra argument to the class, which is allowed by php. So I think this can be a normal required argument.

oleg-andreyev replied:
You're right.

iltar commented on src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php:
```diff
@@ -12,198 +12,331 @@
+        $this->event
+            ->expects($this->any())
+            ->method('getRequest')
+            ->will($this->returnValue($this->request))
```
Not sure if there's a standard defined for this, but you can also write:

```php
$this->event->method('getRequest')->willReturn($this->request);
```

oleg-andreyev replied:
Personally I do prefer shortcuts like `willReturn`, but I decided to stick with style that was before

iltar commented on src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php:
```diff
@@ -12,198 +12,331 @@
+    {
+        parent::__construct($roles);
+        $this->setUser($user);
+        parent::setAuthenticated(count($roles) > 0);
```
Is this really needed? Can just set it to true I suppose. It looks like this logic was duplicated from the UsernamePasswordToken, but doesn't exactly serve a purpose. You can also be authenticated without roles.

oleg-andreyev replied:
Indeed I wanted to mimic `UsernamePasswordToken`, but `UsernamePasswordToken` has a lot of parameters in it's constructor (lazy)

iltar commented:
Have you tested what happens if "always authenticated" is set to true?

oleg-andreyev commented:
Yes, that was the main reason why I reused `AuthorizationChecker`

oleg-andreyev commented:
@nicolas-grekas any chances that this PR could be merged into 2.7?

nicolas-grekas commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -68,4 +77,19 @@ public function handle(GetResponseEvent $event)
+
+        if (null === $attributes) {
+            return;
+        }
```
All the previous lines in the new methods duplicate the logic in "handle".
Would it make sense to not add any newmethod, and instead move the "if" block throwing "AuthenticationCredentialsNotFoundException" after these lines in "handle"?

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php:
```diff
@@ -12,198 +12,331 @@
+    private $authenticationManager;
+    private $authorizationChecker;
+
+    public function setUp()
```
I'm not sure about this change. We prefer not doing refactos of the test suite. Could you revert these changes?

oleg-andreyev replied:
No problem, I'll revert refactoring.

nicolas-grekas commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -59,6 +58,15 @@ public function handle(GetResponseEvent $event)
         }
 
+        if (null !== $this->authorizationChecker) {
+            $this->authorizationChecker->isGranted($attributes, $request);
```
shouldn't this be "if (!...isGranted()) throw..."?

oleg-andreyev replied:
True, I've adjusted the code

nicolas-grekas commented:
LGTM code-style wise. But since I'm not an expert in this component, I'll let others approve the PR.

oleg-andreyev commented:
@nicolas-grekas who has more experience in security component? Got an email with reminder about 2.7 that it won't be maintained expect security fixes

wouterj reviewed (approved):
Looks like a very valid bug fix to me. It makes sense to de-duplicate logic and the described scenario's do appear to be the correct behaviour.

`always_authenticate_before_granting` is disabled by default, so this will only impact projects that explicitly wanted the described behaviour.

:+1: for merging this in 2.7

fabpot commented:
This is more a behavior change than a bug fix to me. I don't think this is something that can be merged in 2.7.

chalasr commented:
Agreed. The bug is in the current documentation, not in the code.

oleg-andreyev commented:
I agree that it's a behavior change, but to me current behavior is unexpected (and not documented) which is caused by inconsistent logic between AccessListener and AuthorizationChecker. 

I usually treat unexpected behavior as bug.

But if majority agrees, I'll rebase changes on master all together will deprecations for AccessListener

fabpot commented:
Drawing the line between a bug and a behavior change is always very difficult. But in this case, I think this is clearly not something we can treat as a bug (especially case 2 which has been reported several times and even if this looks "better" to have the new behavior, this is really a new feature).

wouterj commented:
The inconsistency bit here is what makes this PR very difficult to mark as feature or bug.

For instance:

**Example 1.1**
```yaml
security:
    # ...
    access_control:
        - { pattern: ^/admin, roles: 'ROLE_ADMIN' }
```

**Example 1.2**
```php
/** @Route("/admin") */
public function admin()
{
    $this->denyAccessUnlessGranted('ROLE_ADMIN');
}
```

Before this PR, these two would act differently: (1.1) will not re-authenticate when `alwaysAuthenticated` is true, while (1.2) will.

---

Furthermore, scenario 2 described in the PR description will only act in this way if `always_authenticate` is explicitly set to `true`. A default Symfony app (i.e. most of the app around) would not have different behaviour.

All current authentication managers and providers are already capable of dealing with re-authentication, because example (1.2) is a feature since the very start of Symfony.

---

Given these three points, I think this is not so much a behaviour change as it is about fixing inconsistent behaviour. That's why I would indicate this as a bug fix.

---

Anyway, the core team is set-up to precisily define whether something is a bug or feature. So feel free to merge in whatever branch you like :smiley:

oleg-andreyev commented:
Changed base to 3.4.

nicolas-grekas commented on src/Symfony/Component/Security/Http/Firewall/AccessListener.php:
```diff
@@ -47,10 +50,6 @@ public function __construct(TokenStorageInterface $tokenStorage, AccessDecisionM
      */
     public function handle(GetResponseEvent $event)
     {
-        if (null === $token = $this->tokenStorage->getToken()) {
```
I think we should keep this block. The exception message is more tailored than the one in `AuthorizationChecker`

oleg-andreyev replied:
In my opinion message from `AuthorizationChecker` is better from DX perspective, because it is hinting about why token may be missing.

nicolas-grekas replied:
The message in `AuthorizationChecker` reads as

> The token storage contains no authentication token. One possible reason may be that there is no firewall configured for this URL.

Is it possible to have the AccessListener configured but not firewall configured? If not, then the message would be misleading.

But there is something else: this moves the check for a token after the check for attributes. Is it legitimate?

oleg-andreyev replied:
`AccessListener` is configured by https://github.com/symfony/symfony/blob/5e28ac31504e3a7cd2d16b0216182d767f37ad41/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L169

So it's impossible to have configured `AccessListener` without firewalls, unless developers are doing it somehow else.

nicolas-grekas replied:
OK to me, lets continue here :)

nicolas-grekas reviewed (changes requested):
This is staling because changing the behavior on lower branches poses a risk to break BC for someone.
Let's merge on master. If there is anything to warn about in UPGRADE* file and CHANGELOG of the component, it should be added also.
OK to move forward with this plan @oleg-andreyev?

oleg-andreyev committed:
#27121 reusing AuthorizationChecker into AccessListener
- added deprecation notice about old-signature
- added deprecation annotation to properties
- added legacy group to old unit tests
- added new unit tests
- adjusted security.access_listener

xabbuh commented:
If some of the proposed changes are a bugfix, we need to split this PR: Move everything that is not about the bugfix to a new PR targetting `master` and then rebase the changes for the bugfix onto the `3.4` branch.

oleg-andreyev commented:
This PR got messy, I'm closing it and will provide a cleaner version.
